File: golang/tests/pkg/integration/lifecycle_policy_integration_test.go
package integration

import (
	"os"
	"strings"
	"testing"

	"github.com/aliyun/wuying-agentbay-sdk/golang/pkg/agentbay"
)

func TestLifecyclePolicyCustom(t *testing.T) {
	apiKey := os.Getenv("AGENTBAY_API_KEY")
	if apiKey == "" {
		t.Skip("AGENTBAY_API_KEY not set")
	}
	client, err := agentbay.NewAgentBay(apiKey, nil)
	if err != nil {
		t.Fatalf("failed to create AgentBay: %v", err)
	}
	lp := agentbay.NewLifecyclePolicy()
	lp.IdleReleaseTimeout = 10
	lp.MaxRuntime = 60
	base := agentbay.NewCreateSessionParams().
		WithImageId("linux_latest").
		WithLabels(map[string]string{
			"test": "lifecycle-policy",
			"sdk":  "golang",
			"case": "custom",
		})
	params, err := base.WithLifecyclePolicy(lp)
	if err != nil {
		t.Fatalf("WithLifecyclePolicy: %v", err)
	}
	result, err := client.Create(params)
	if err != nil {
		t.Fatalf("Create: %v", err)
	}
	if result.Session == nil || strings.TrimSpace(result.Session.SessionID) == "" {
		t.Fatal("session not created")
	}
	s := result.Session
	defer func() {
		if s != nil {
			_, _ = s.Delete()
		}
	}()
	cmdResult, err := s.Command.ExecuteCommand("echo hello")
	if err != nil {
		t.Fatalf("ExecuteCommand: %v", err)
	}
	out := strings.TrimSpace(cmdResult.Output)
	if out != "hello" {
		t.Fatalf("unexpected output %q want %q", cmdResult.Output, "hello")
	}
}

func TestLifecyclePolicyManualRelease(t *testing.T) {
	apiKey := os.Getenv("AGENTBAY_API_KEY")
	if apiKey == "" {
		t.Skip("AGENTBAY_API_KEY not set")
	}
	client, err := agentbay.NewAgentBay(apiKey, nil)
	if err != nil {
		t.Fatalf("failed to create AgentBay: %v", err)
	}
	lp := agentbay.NewLifecyclePolicyManualRelease()
	base := agentbay.NewCreateSessionParams().
		WithImageId("linux_latest").
		WithLabels(map[string]string{
			"test": "lifecycle-policy",
			"sdk":  "golang",
			"case": "manual",
		})
	params, err := base.WithLifecyclePolicy(lp)
	if err != nil {
		t.Fatalf("WithLifecyclePolicy: %v", err)
	}
	result, err := client.Create(params)
	if err != nil {
		t.Fatalf("Create: %v", err)
	}
	if result.Session == nil || strings.TrimSpace(result.Session.SessionID) == "" {
		t.Fatal("session not created")
	}
	s := result.Session
	defer func() {
		if s != nil {
			_, _ = s.Delete()
		}
	}()
	cmdResult, err := s.Command.ExecuteCommand("echo manual")
	if err != nil {
		t.Fatalf("ExecuteCommand: %v", err)
	}
	out := strings.TrimSpace(cmdResult.Output)
	if out != "manual" {
		t.Fatalf("unexpected output %q want %q", cmdResult.Output, "manual")
	}
}

func TestLifecyclePolicyDefault(t *testing.T) {
	apiKey := os.Getenv("AGENTBAY_API_KEY")
	if apiKey == "" {
		t.Skip("AGENTBAY_API_KEY not set")
	}
	client, err := agentbay.NewAgentBay(apiKey, nil)
	if err != nil {
		t.Fatalf("failed to create AgentBay: %v", err)
	}
	lp := agentbay.NewLifecyclePolicy()
	base := agentbay.NewCreateSessionParams().
		WithImageId("linux_latest").
		WithLabels(map[string]string{
			"test": "lifecycle-policy",
			"sdk":  "golang",
			"case": "default",
		})
	params, err := base.WithLifecyclePolicy(lp)
	if err != nil {
		t.Fatalf("WithLifecyclePolicy: %v", err)
	}
	result, err := client.Create(params)
	if err != nil {
		t.Fatalf("Create: %v", err)
	}
	if result.Session == nil || strings.TrimSpace(result.Session.SessionID) == "" {
		t.Fatal("session not created")
	}
	s := result.Session
	defer func() {
		if s != nil {
			_, _ = s.Delete()
		}
	}()
}

File: typescript/tests/integration/lifecycle-policy.integration.test.ts
import { AgentBay, CreateSessionParams, LifecyclePolicy } from "../../src";
import { getTestApiKey } from "../utils/test-helpers";

describe("LifecyclePolicy integration", () => {
  const imageId = "linux_latest";

  it("should create session with custom lifecycle policy", async () => {
    const agentBay = new AgentBay({ apiKey: getTestApiKey() });
    const policy = new LifecyclePolicy({
      idleReleaseTimeout: 10,
      maxRuntime: 60,
    });
    const params: CreateSessionParams = {
      imageId,
      labels: {
        test: "lifecycle-policy",
        sdk: "typescript",
        case: "custom",
      },
      lifecyclePolicy: policy,
    };
    const result = await agentBay.create(params);
    expect(result.success).toBe(true);
    expect(result.session).toBeDefined();
    const session = result.session!;
    try {
      const cmdResult = await session.command.executeCommand("echo hello");
      expect(cmdResult.output?.trim()).toBe("hello");
    } finally {
      await agentBay.delete(session);
    }
  }, 120000);

  it("should create session with manual release", async () => {
    const agentBay = new AgentBay({ apiKey: getTestApiKey() });
    const policy = new LifecyclePolicy({ manualRelease: true });
    const params: CreateSessionParams = {
      imageId,
      labels: {
        test: "lifecycle-policy",
        sdk: "typescript",
        case: "manual",
      },
      lifecyclePolicy: policy,
    };
    const result = await agentBay.create(params);
    expect(result.success).toBe(true);
    expect(result.session).toBeDefined();
    const session = result.session!;
    try {
      const cmdResult = await session.command.executeCommand("echo manual");
      expect(cmdResult.output?.trim()).toBe("manual");
    } finally {
      await agentBay.delete(session);
    }
  }, 120000);

  it("should create session with default lifecycle policy", async () => {
    const agentBay = new AgentBay({ apiKey: getTestApiKey() });
    const params: CreateSessionParams = {
      imageId,
      labels: {
        test: "lifecycle-policy",
        sdk: "typescript",
        case: "default",
      },
      lifecyclePolicy: new LifecyclePolicy(),
    };
    const result = await agentBay.create(params);
    expect(result.success).toBe(true);
    expect(result.session).toBeDefined();
    const session = result.session!;
    await agentBay.delete(session);
  }, 120000);
});
